dev/src/dev/migration.clj

```diff
@@ -0,0 +1,62 @@
+(ns dev.migration
+  (:require
+   [metabase.db :as mdb]
+   [metabase.db.liquibase :as liquibase]
+   [toucan2.core :as t2])
+  (:import
+   (liquibase Liquibase)))
+
+(set! *warn-on-reflection* true)
+
+(defn- latest-migration
+  []
+  ((juxt :id :comments)
+   (t2/query-one {:select [:id :comments]
+                  :from   [:databasechangelog]
+                  :order-by [[:orderexecuted :desc]]
+                  :limit 1})))
+(defn migrate!
+  "Run migrations for the Metabase application database. Possible directions are `:up` (default), `:force`, `:down`, and
+  `:release-locks`. When migrating `:down` pass along a version to migrate to (44+)."
+  ([]
+   (migrate! :up))
+  ;; do we really use this in dev?
+  ([direction & [version]]
+   (mdb/migrate! (mdb/db-type) (mdb/data-source)
+                 direction version)
+   (println "Latest migration:" (latest-migration))))
+
+(defn- rollback-n-migration!
+  [n]
+  (with-open [conn (.getConnection (mdb/data-source))]
+    (liquibase/with-liquibase [^Liquibase liquibase conn]
+      (liquibase/with-scope-locked liquibase
+        (.rollback liquibase n "")))))
+
+(defn- id->n-to-rollback
+  [id]
+  (->> (t2/query-one {:select [[:%count.* :count]]
+                              :from   [:databasechangelog]
+                              :where  [:> :orderexecuted {:select [:orderexecuted]
+                                                          :from   [:databasechangelog]
+                                                          :where  [:= :id id]
+                                                          :limit 1}]
+                              :limit 1})
+       :count
+       ;; includes the selected id
+       inc))
```

Comment:
I'm a bit confused by why this is needed, would we want to rollback until the given migration was the last one applied, right? That's what I'd expected, if I'm remembering my long lost Rails days clearly.

Even if it is needed, wouldn't it be better to use `:>=` in the `:where` clause instead of incrementing afterwards?

Reply:
OK, my memory actually does fail me. Rails also has "inclusive" semantics for their `db:rollback` command, and I was thinking about the semantics for `db:migrate`.

I we're going with the kwargs interface though, I still expect `:to` to mean exclusive, i.e. ending on that migration. I'm not sure what a good preposition or noun for inclusive would be, perhaps `:through`?

Reply:
> I'm a bit confused by why this is needed, would we want to rollback until the given migration was the last one applied, right? 

For me the usual way would be to "rollback all my migrations", so I'd copy ids of my migrations and delete them from the `databasechangelog`. So to me this seems like a good workflow?

Don't care about Rails. 😁

Reply:
Rails cares about you though, and does what you want 😆